sky/templates/kubernetes-port-forward-proxy-command.sh.j2

```diff
@@ -13,31 +13,42 @@ if ! command -v lsof > /dev/null; then
   exit 1
 fi
 
-# Function to check if port is in use
-is_port_in_use() {
-    local port="$1"
-    lsof -i :${port} > /dev/null 2>&1
-}
-
-# Start from a fixed local port and increment if in use
-local_port={{ local_port }}
-while is_port_in_use "${local_port}"; do
-    local_port=$((local_port + 1))
+# Establishes connection between local port and the ssh jump pod using kube port-forward
+# Instead of specifying a port, we let kubectl select a random port.
+# This is preferred because of socket re-use issues in kubectl port-forward,
+# see - https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-769185879
+KUBECTL_OUTPUT=$(mktemp)
+kubectl port-forward svc/{{ ssh_jump_name }} :22 > "${KUBECTL_OUTPUT}" 2>&1 &
+
+# Capture the PID for the backgrounded kubectl command
+K8S_PORT_FWD_PID=$!
+
+# Wait until kubectl port-forward is ready
+while ! grep -q "Forwarding from 127.0.0.1" "${KUBECTL_OUTPUT}"; do
+    sleep 0.1
 done
 
-# Establishes connection between local port and the ssh jump pod
-kubectl port-forward svc/{{ ssh_jump_name }} "${local_port}":22 &
+# Extract the local port number assigned by kubectl.
+local_port=$(awk -F: '/Forwarding from 127.0.0.1/{print $2}' "${KUBECTL_OUTPUT}" | awk -F' ' '{print $1}')
 
-# Terminate the port-forward process when this script exits.
+# Clean up the temporary file
+rm -f "${KUBECTL_OUTPUT}"
+
+# Add handler to terminate the port-forward process when this script exits.
 K8S_PORT_FWD_PID=$!
-trap "kill $K8S_PORT_FWD_PID" EXIT
+trap "[[ -e /proc/$K8S_PORT_FWD_PID ]] && kill $K8S_PORT_FWD_PID" EXIT
 
-# checks if a connection to local_port of 127.0.0.1:[local_port] is established
+# Checks if a connection to local_port of 127.0.0.1:[local_port] is established
 while ! nc -z 127.0.0.1 "${local_port}"; do
     sleep 0.1
 done
 
+# To avoid errors when many concurrent requests are sent (see https://github.com/skypilot-org/skypilot/issues/2628),
+# we add a random delay before establishing the socat connection.
+# Empirically, this needs to be at least 1 second. We set this to be random between 1 and 2 seconds.
+sleep $(shuf -i 10-20 -n 1 | awk '{printf "%f", $1/10}')
+
 # Establishes two directional byte streams to handle stdin/stdout between
 # terminal and the jump pod.
 # socat process terminates when port-forward terminates.
-socat - tcp:127.0.0.1:"${local_port}"
\ No newline at end of file
+socat - tcp:127.0.0.1:"${local_port}"
```

Comment:
Also it would be nice to have a newline at the EOF.

Reply:
Unfortunately, the `socat` command succeeds and starts running without any error even though the underlying `kubectl port-forward` may not be functional. Since there's no signal that I can use to detect if kubectl port-forward is ready, I used random wait which has been sufficient in my testing.

Reply:
Ahh, makes sense. I think it should be fine to wait for a random gap. Just curious, something that checks the socket connection would work for checking if the connection gets setup correctly, like the following:
https://github.com/skypilot-org/skypilot/blob/029f886a31953a7b544ad4956ab48cf85cd02fa6/sky/provision/provisioner.py#L239-L258

Reply:
nvm, it seems we have already done the `nc` check before this wait. Waiting looks good to me.